crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -127,31 +129,191 @@ impl Extension for QueryLimitsChecker {
         let cfg = ctx
             .data::<ServiceConfig>()
             .expect("No service config provided in schema data");
-        if query.len() > cfg.limits.max_query_payload_size as usize {
-            metrics
-                .request_metrics
-                .query_payload_too_large_size
-                .observe(query.len() as f64);
-            info!(
-                query_id = %query_id,
-                session_id = %session_id,
-                error_code = code::BAD_USER_INPUT,
-                "Query payload is too large: {}",
-                query.len()
-            );
-
-            return Err(graphql_error(
-                code::BAD_USER_INPUT,
-                format!(
-                    "Query payload is too large. The maximum allowed is {} bytes",
-                    cfg.limits.max_query_payload_size
-                ),
-            ));
-        }
-
         // Document layout of the query
         let doc = next.run(ctx, query, variables).await?;
 
+        let mutation_node = doc
+            .operations
+            .iter()
+            .find(|(_, operation)| operation.node.ty == OperationType::Mutation);
+
+        // this is a map of type node position, (tx_bytes, signatures) bytes for all nodes of type
+        // executeTransactionBlock
+        // let mut tx_bytes_sigs = BTreeMap::new();
+        let mut tx_bytes_sigs_len = 0;
+
+        let query_len = query.len();
+        let variables_len = variables.to_string().len();
```

Comment:
I would really like to avoid using the `to_string` implementation here to calculate the length.  I looked at [how it is implemented](https://docs.rs/async-graphql-value/7.0.6/src/async_graphql_value/variables.rs.html#16), and it is intended to be used as a human-readable representation of the variables.

Its size is currently proportional to the size of the content of the variables part of the request payload but it's going to ignore punctuation (quotes around names) and whitespace, and crucially there is no guarantee that the `Display` implementation will stay like this -- it could change to include its own indentation/punctuation, perform truncation, etc, all of which will impact our limits logic in a way that's difficult to detect. As a user, you will also see a certain size usage reported and have no idea how it was arrived at because it went through some opaque process.

Here's a suggested alternative, let me know what you think:

- Capture the overall request size outside of the GraphQL service (e.g. from Axum) and perform a high level check at that point that the overall size doesn't exceed the combined transaction payload and read limit. 
  - If it does, complain right there, 
  - otherwise stash the payload size in the GraphQL context for the query limit checker to pick up.
- In the limits checker, iterate over the query structure, looking for `executeTransactionBlock` and `dryRunTransactionBlock` elements, and discount their transaction payloads from the overall size.
  - You may also be able to drop the test for the operation type, because we are looking for both executions and dry-runs. 
- Check at the end that the remaining overall size is within the query payload limit.

The logic related to discounting payload sizes is tricky when it comes to variables (I think the current implementation falls into the trap of the first bullet below -- double counting), here's the best I could come up with but let me know if you can think of something simpler.
  - Need to be careful not to double count payloads that are stored in variables and used multiple times.
  - Similarly we need to be careful not to discount a variable that is used in both a transaction payload and a read.
  - So this means we can discount (once) a variable that is used exclusively in dry runs and transactions.

Reply:
Thanks for sharing your thoughts. Valid concerns around using `to_string` impl.
Revising now.

Reply:
Regarding variables, this is what I came up with:
- only take the variables from `dryRun` and `executeTransactionBlock`, and keep track of them (in case one is used in multiple places). 
- subtract the size of variables from above from content length. 
- if content length > max query payload, then error. 

For each variable/txBytes(+signatures in case of executeTxBlock), it checks if the payload_transaction is less than max transaction payload size.